### PR TITLE
Demand - Proof of Concept

### PR DESCRIPTION
**Pull request by agritheory**
This feature aligns in part with cases 2 and 3 of [this comment](https://github.com/agritheory/beam/issues/15#issuecomment-2061985871) of #15. The ideas here need to be refined in order to better understand how it should be used, which is why I'm starting a pull request. As a feature, I think it is likely a prerequisite for performant mobile views, which is why I was developing it in the context of this app, but as a feature, it more strongly aligns with Inventory Tools. We may want to finish the work there and then import it here as a dependency.

The purpose of this feature is to be able to quickly determine if a Work Order or Delivery Note can be started, based on material availability in the required location(s).

I tried several iterations of trying to organize this data in Redis before ultimately giving up - Redis' capabilities for storing structured data is limited and the inability to search by value means that filtering is only possible with the construction a several hash maps, which serve as unsynchronized indexes to other hash maps containing serialized data. It just felt like the wrong tool. 

Refactoring to sqllite meant that I could preserve the tabular nature of the data, add indexes and use an ultimately more familiar approach that is still essentially in-memory. Reads from sqllite are notoriously fast and the write operations should be trivial. We may have to implement a write lock or [ACID files](https://pypi.org/project/acidfile/) approach to make sure the file isn't corrupted in a case where it would be manipulated by more than one user.

<hr>

To be considered feature complete, I think the following has to be added:

- [x] Add nuance to the warehouse(s) signaling demand. This should be configurable and based on a multiselect of warehouse type and parent(s) so as to exclude quality/ quarantine warehouses, subcontractor warehouses and/or third-party warehousing solutions. #107 
- [x] Test suite should integrate this (leveraging `pytest-order`) such that the demand result is re-examined after each stock transaction. This also likely means doing a complete manufacturing cycle in the test suite so that we can track manufactured goods all the way to Delivery Note. #108
- [ ] Create endpoints with filters for `company`, `item_code`, `warehouse`, `workstation` and `assigned` so that this can integrate with #21 
- [x] I've only implemented the bare minimum of updates to the demand table and have not accounted for any logic on how rows should be removed when demand is satisfied. #108
- [ ] This feature should be defeatable in a settings document, probably per-company
- [x] To avoid file corruption concerns, some ACID compliant write wrapper should be added #109 
- [ ] #110 
- [x] #111
- [ ] It would be good to add a "planned availability" feature #157
- [ ] This should be adaptive to Inventory Dimensions, batch and/or serial numbers #168

### CODE REVIEW

File: beam/tests/setup.py [hunk 1]
```diff
@@ -1,577 +1,588 @@
...
+		)
+		if i.is_purchase_item and item.get("supplier"):
+			i.append("supplier_items", {"supplier": item.get("supplier")})
+		if i.item_code == "Parchment Paper":
+			i.append("uoms", {"uom": "Box", "conversion_factor": 100})
+			i.purchase_uom = "Box"
+		if i.item_code in ("Water", "Ice Water"):
+			i.append("uoms", {"uom": "Gallon Liquid (US)", "conversion_factor": 15.142})
+			i.purchase_uom = "Gallon Liquid (US)"
+			i.valuation_rate = 0.01 if i.item_code == "Water" else 0.02
```
Comment (Alchez) [inline beam/tests/setup.py]: @agritheory there was some conflict in the HU tests, but I wasn't sure which changes to include, so I've added both sets of changes. Can you have a look please?

File: beam/tests/setup.py [hunk 2]
```diff
@@ -1,577 +1,588 @@
...
+	water.stock_entry_type = water.purpose = "Material Receipt"
+	water.append(
+		"items",
+		{
+			"item_code": "Water",
+			"qty": 1000,
+			"t_warehouse": "Refrigerator - APC",
+			"uom": "Gallon Liquid (US)",
+			"basic_rate": 0.15,
+			"expense_account": "5111 - Cost of Goods Sold - APC",
```
Comment (Alchez) [inline beam/tests/setup.py]: @agritheory this is included in the changes I mentioned above

File: beam/tests/setup.py [hunk 3]
```diff
@@ -1,577 +1,588 @@
...
+	)
+	water.append(
+		"items",
+		{
+			"item_code": "Ice Water",
+			"qty": 1000,
+			"t_warehouse": "Refrigerator - APC",
+			"basic_rate": 0.30,
+			"expense_account": "5111 - Cost of Goods Sold - APC",
+		},
```
Comment (Alchez) [inline beam/tests/setup.py]: @agritheory this is included in the changes I mentioned above

File: beam/tests/setup.py [hunk 4]
```diff
@@ -1,577 +1,588 @@
...
+		)
+		pp.get_mr_items()
+	for item in pp.po_items:
+		item.planned_start_date = settings.day
+	pp.get_sub_assembly_items()
+	start_time = datetime.datetime(settings.day.year, settings.day.month, settings.day.day, 0, 0)
+	for item in pp.sub_assembly_items:
+		item.schedule_date = start_time
+		time = frappe.get_value("BOM Operation", {"parent": item.bom_no}, "sum(time_in_mins) AS time")
+		start_time += datetime.timedelta(minutes=time + 2)
```
Comment (Alchez) [inline beam/tests/setup.py]: @agritheory this is included in the changes I mentioned above

File: beam/beam/demand/demand.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+		row.key = frappe.generate_hash()
+		row.delivery_date = str(calendar.timegm(get_datetime(row.delivery_date).timetuple()))
+		row.net_required_qty = str(row.net_required_qty)
+		row.actual_qty = str(get_balance_qty_from_sle(row.item_code, row.warehouse))
+		row.indent = str(0)
+		output.append(row)
+
+		if frappe.db.get_value("Warehouse", row.warehouse, "is_group"):
+			warehouses_to_check = [row.warehouse]
+			warehouses_to_check.extend(get_descendant_warehouses("Ambrosia Pie Company", row.warehouse))
```
Comment (agritheory) [inline beam/beam/demand/demand.py]: String literal here

File: beam/beam/demand/demand.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+if TYPE_CHECKING:
+	from erpnext.accounts.doctype.purchase_invoice.purchase_invoice import PurchaseInvoice
+	from erpnext.accounts.doctype.sales_invoice.sales_invoice import SalesInvoice
+	from erpnext.stock.doctype.delivery_note.delivery_note import DeliveryNote
+	from erpnext.stock.doctype.purchase_receipt.purchase_receipt import PurchaseReceipt
+	from erpnext.stock.doctype.stock_entry.stock_entry import StockEntry
+	from erpnext.stock.doctype.stock_reconciliation.stock_reconciliation import StockReconciliation
+
+
+def build_demand_map() -> None:
```
Comment (agritheory) [inline beam/beam/demand/demand.py]: Add company argument

**Comment by agritheory** [comment]
@Alchez Can you re-implement the demand action map you made inside of `build_demand_map`? Using the Demand Map report, you can see that transfers and updates aren't implemented. 

The tests are currently passing for me locally, but that's because they're not really testing this enough - the results aren't correct.

**Comment by Alchez** [comment]
> @Alchez Can you re-implement the demand action map you made inside of `build_demand_map`? Using the Demand Map report, you can see that transfers and updates aren't implemented.
> 
> The tests are currently passing for me locally, but that's because they're not really testing this enough - the results aren't correct.

@agritheory, I've created #145 since I wasn't sure about the implementation of the demand maps. We can discuss the map-specific changes there.

**Comment by agritheory** [comment]
Merging features as-is into mobile_v15, but leaving this PR open for completeness

**Pull request merged**